peas/tests/serial_handlers/protocol_arduino_simulator.py

```diff
@@ -0,0 +1,422 @@
+
+        Raises:
+            SerialTimeoutException: In case a write timeout is configured for
+                the port and the time is exceeded.
+        """
+        if not isinstance(data, (bytes, bytearray)):
+            raise ValueError("write takes bytes")
+        data = bytes(data)  # Make sure it can't change.
+        self.logger.info('FakeArduinoSerialHandler.write({!r})', data)
+        count = len(data)
+        try:
+            self.relay_queue.put(data, block=True, timeout=self.write_timeout)
+            return len(data)
+        except queue.Full:
+            # This exception is "lossy" in that the caller can't tell how much was written.
+            raise serialutil.writeTimeoutError
+
+    # --------------------------------------------------------------------------
+
+    @property
+    def is_config_ok(self):
+        """Does the caller ask for the correct serial device config?"""
+        # The default Arduino data, parity and stop bits are: 8 data bits, no parity, one stop bit.
+        return (self.baudrate == 9600 and self.bytesize == serialutil.EIGHTBITS and
+                self.parity == serialutil.PARITY_NONE and not self.rtscts and not self.dsrdtr)
+
+    def _read1(self, timeout_obj):
+        """Read 1 byte of input, of type bytes."""
+        if not self.is_open:
+            raise serialutil.portNotOpenError
+        if not self.json_bytes:
+            try:
+                entry = self.json_queue.get(block=True, timeout=timeout_obj.time_left())
+                assert isinstance(entry, bytes)
+                self.json_bytes.extend(entry)
+            except queue.Empty:
+                return None
+        if not self.json_bytes:
+            return None
+        c = bytes(self.json_bytes[0:1])
+        del self.json_bytes[0:1]
+        return c
+
+    # --------------------------------------------------------------------------
+    # There are a number of methods called by SerialBase that need to be
+    # implemented by sub-classes, assuming their calls haven't been blocked
+    # by replacing the calling methods/properties. These are no-op
+    # implementations.
+
+    def _reconfigure_port(self):
+        """Reconfigure the open port after a property has been changed.
+
+        If you need to know which property has been changed, override the
+        setter for the appropriate properties.
+        """
+        need_thread = self.is_open and self.is_config_ok
+        if need_thread and not self.simulator_thread:
+            _drain_queue(self.relay_queue)
+            _drain_queue(self.json_queue)
+            self.json_bytes.clear()
+            self.stop.clear()
+            self.simulator_thread = threading.Thread(
+                name='Device Simulator', target=lambda: self.device_simulator.run())
+            self.simulator_thread.start()
+        elif self.simulator_thread and not need_thread:
+            self.stop.set()
+            self.simulator_thread.join(timeout=30.0)
```

Comment:
Do we need this long of a timeout here?

Reply:
It shouldn't get that long, unless you're running in the debugger. But, again, this is intended to be used from tests. I don't want to make it so short that on an underpowered or overloaded system (like Travis) we start to get flakey results.

Reply:
Sounds good.